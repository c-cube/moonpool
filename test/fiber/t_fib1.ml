open Moonpool
module F = Moonpool_fib.Fiber

let ( let@ ) = ( @@ )
let runner = Ws_pool.create ~num_threads:8 ()
let lock = Lock.create ()

let logf fmt =
  Printf.ksprintf
    (fun s ->
      let out = stdout in
      (let@ () = Lock.with_ lock in
       output_string out s);
      flush out)
    fmt

let () =
  logf "============\nstart\n";
  let fib =
    F.spawn_top ~on:runner @@ fun () ->
    let subs =
      List.init 5 (fun i ->
          F.spawn_link ~protect:false @@ fun () ->
          Thread.delay 0.000_01;
          i)
    in

    ignore
      (F.spawn_link ~protect:false @@ fun () ->
       Thread.delay 0.2;
       logf "other fib done\n%!"
        : _ F.t);
    logf "wait for subs\n%!";
    List.iteri
      (fun i f ->
        logf "await fiber %d\n%!" i;
        logf "cur fiber is some: %b\n%!"
          (Option.is_some @@ F.Private_.get_cur ());
        let res = F.await f in
        logf "cur fiber is some: %b\n%!"
          (Option.is_some @@ F.Private_.get_cur ());
        Thread.delay 0.000_01;
        F.yield ();
        logf "res %d = %d\n%!" i res)
      subs;
    logf "main fiber done\n%!"
  in

  Fut.wait_block_exn @@ F.res fib;
  logf "main fiber exited\n%!";
  ()

let () =
  (* same but now, cancel *)
  logf "============\nstart\n";
  let fib =
    F.spawn_top ~on:runner @@ fun () ->
    let@ () =
      F.with_self_cancel_callback (fun ebt ->
          logf "main fiber cancelled with %s\n%!" @@ Exn_bt.show ebt)
    in

    let subs =
      List.init 10 (fun i ->
          F.spawn_link ~protect:false @@ fun () ->
          let@ () =
            F.with_self_cancel_callback (fun _ ->
                logf "sub-fiber %d was cancelled\n%!" i)
          in
          Thread.delay (float i *. 0.001);
          F.yield ();
          if i = 7 then (
            logf "I'm fiber %d and I'm about toâ€¦\n%!" i;
            failwith "oh no!"
          );
          i)
    in

    List.iteri
      (fun i fib ->
        F.on_result fib (function
          | Ok _ -> logf "fiber %d resolved as ok\n%!" i
          | Error _ -> logf "fiber %d resolved as error\n%!" i))
      subs;

    ignore
      (F.spawn_link ~protect:false @@ fun () ->
       Thread.delay 0.2;
       logf "other fib done\n%!"
        : _ F.t);

    logf "wait for subs\n%!";
    List.iteri
      (fun i f ->
        logf "await fiber %d\n%!" i;
        let res = F.await f in
        logf "res %d = %d\n%!" i res)
      subs;
    logf "yield\n%!";
    F.yield ();
    logf "yielded\n%!";
    logf "main fiber done\n%!"
  in

  F.on_result fib (function
    | Ok () -> logf "main fiber result: ok\n%!"
    | Error ebt -> logf "main fiber result: error %s\n%!" (Exn_bt.show ebt));

  (try Fut.wait_block_exn @@ F.res fib
   with Failure msg -> logf "main fib failed with %S\n%!" msg);
  logf "main fiber exited\n%!";
  ()
