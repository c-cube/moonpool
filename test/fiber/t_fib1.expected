============
start
wait for subs
await fiber 0
cur fiber is some: true
cur fiber is some: true
res 0 = 0
await fiber 1
cur fiber is some: true
cur fiber is some: true
res 1 = 1
await fiber 2
cur fiber is some: true
cur fiber is some: true
res 2 = 2
await fiber 3
cur fiber is some: true
cur fiber is some: true
res 3 = 3
await fiber 4
cur fiber is some: true
cur fiber is some: true
res 4 = 4
main fiber done
other fib done
main fiber exited
============
start
wait for subs
await fiber 0
fiber 0 resolved as ok
res 0 = 0
await fiber 1
fiber 1 resolved as ok
res 1 = 1
await fiber 2
fiber 2 resolved as ok
res 2 = 2
await fiber 3
fiber 3 resolved as ok
res 3 = 3
await fiber 4
fiber 4 resolved as ok
res 4 = 4
await fiber 5
fiber 5 resolved as ok
res 5 = 5
await fiber 6
res 6 = 6
await fiber 7
fiber 6 resolved as ok
I'm fiber 7 and I'm about toâ€¦
sub-fiber 8 was cancelled
fiber 8 resolved as error
sub-fiber 9 was cancelled
fiber 9 resolved as error
main fiber result: error Failure("oh no!")
fiber 7 resolved as error
main fib failed with "oh no!"
main fiber exited
